src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs

```diff
@@ -22,6 +24,121 @@ internal sealed partial class SyntaxTreeIndex : AbstractPersistableState, IObjec
         private static readonly ConditionalWeakTable<BranchId, ConditionalWeakTable<DocumentId, SyntaxTreeIndex>> s_cache =
             new ConditionalWeakTable<BranchId, ConditionalWeakTable<DocumentId, SyntaxTreeIndex>>();
 
+        public readonly VersionStamp Version;
+
+        private void WriteVersion(ObjectWriter writer, string formatVersion)
+        {
+            writer.WriteString(formatVersion);
+            this.Version.WriteTo(writer);
+        }
+
+        private static bool TryReadVersion(ObjectReader reader, string formatVersion, out VersionStamp version)
+        {
+            version = VersionStamp.Default;
+            if (reader.ReadString() != formatVersion)
+            {
+                return false;
+            }
+
+            version = VersionStamp.ReadFrom(reader);
+            return true;
+        }
+
+        private static async Task<SyntaxTreeIndex> LoadAsync(
+            Document document, string persistenceName, string formatVersion,
+            Func<ObjectReader, VersionStamp, SyntaxTreeIndex> readFrom, CancellationToken cancellationToken)
+        {
+            var persistentStorageService = document.Project.Solution.Workspace.Services.GetService<IPersistentStorageService>();
+            var syntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
+
+            try
+            {
+                // attempt to load from persisted state
+                using (var storage = persistentStorageService.GetStorage(document.Project.Solution))
+                using (var stream = await storage.ReadStreamAsync(document, persistenceName, cancellationToken).ConfigureAwait(false))
+                {
+                    if (stream == null)
+                    {
+                        return null;
+                    }
+
+                    using (var reader = new StreamObjectReader(stream))
+                    {
+                        if (TryReadVersion(reader, formatVersion, out var persistVersion) &&
+                            document.CanReusePersistedSyntaxTreeVersion(syntaxVersion, persistVersion))
+                        {
+                            return readFrom(reader, syntaxVersion);
+                        }
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                // Storage APIs can throw arbitrary exceptions.
```

Comment:
Should we say something like "if this isn't an IO exception", non-fatal? I'm guessing you've already done this thought experiment, but a comment here might be useful.